### PR TITLE
fix(vue-app, vue-renderer, utils):  respect `trailingSlash` setting for payloads 

### LINKED ISSUE #8476: trailingSlash:true corrupts payloads
### Versions

- nuxt: 2.14.11
- node: 12.18.2

### Steps to reproduce

I'm using `generate` with `{ target: 'static' }` in my config to get a fully-static website.

I'm using `asyncData` to retrieve content from Prismic API.

And I'm using `{ router: { trailingSlash: true } }` in Nuxt config to add trailing slash all URLs.

### What is actually happening?

Problem is, the generated payload files don't have a trailing slash in their keys, so pages's data remains blank, and new calls to API are made on each route change.

### LINKED ISSUE #8488: catchall routes with `trailingSlash: true` lead to 404s when generated
### Reproduction

https://github.com/danielroe/trailingslash-payload-repro

### Steps to reproduce

`yarn generate`

### What is Expected?

`index.html` should be generated

### What is actually happening?

`ERROR   Error generating route "/": This page could not be found`

### PR DESCRIPTION
**Pull request by danielroe**
## Types of changes
- [x] Bug fix (a non-breaking change which fixes an issue)

## Description
This respects the route URL so payloads won't be refreshed when `trailingSlash: true`. Note this will still refresh payloads if the *link* doesn't respect the `trailingSlash` settings - there is still a need for a `<NuxtLink>` that respects the trailingSlash settings.

closes #8488
closes #8476 

## Checklist:

- [ ] ~~I have added tests to cover my changes~~ (will add tests in #8423)
- [x] All new and existing tests are passing.

### CODE REVIEW

**Review by pi0 — DISMISSED** [review]
Notes for later:
- We can expose `withoutQuery`  from ufo later on (or simply using new ponyfill `.pathname`
- We may add tests for full-static to trailing-slash PR

File: packages/vue-app/template/App.js [hunk 1]
```diff
@@ -296,19 +297,23 @@ export default {
...
     getRoutePath(route = '/') {
       const base = this.getRouterBase()
       if (base && route.startsWith(base)) {
         route = route.substr(base.length)
       }
-      return (route.replace(/\/+$/, '') || '/').split('?')[0].split('#')[0]
+      let path = route.split('?')[0].split('#')[0]
+      <% if (!nuxtOptions.router.trailingSlash) { %>
+        path = withoutTrailingSlash(path)
+      <% } %>
```
Comment (danielroe) [inline packages/vue-app/template/App.js:310]: We could normalise to no trailing slashes, which would also solve this problem, but would have the downside of breaking the distinction between parent/child routes (see #8423 - thoughts welcome @pi0).

File: packages/vue-app/template/App.js [hunk 2]
```diff
@@ -296,19 +297,23 @@ export default {
...
-      return (this.$router.options.base || '').replace(/\/+$/, '')
+      return withoutTrailingSlash(this.$router.options.base)
     },
     getRoutePath(route = '/') {
       const base = this.getRouterBase()
       if (base && route.startsWith(base)) {
         route = route.substr(base.length)
       }
-      return (route.replace(/\/+$/, '') || '/').split('?')[0].split('#')[0]
+      let path = route.split('?')[0].split('#')[0]
```
Comment (Atinux) [inline packages/vue-app/template/App.js:307]: Could we expose à utiliser from ufo for this @pi0 ?

Comment (pi0) [reply packages/vue-app/template/App.js:307]: yes. see comment above 😁

Comment (pi0) [reply packages/vue-app/template/App.js:307]: FYI it is possible with latest version of ufo to do `parsePath(route).pathname`

**Comment by pi0** [comment]
Should be available shortly with `nuxt-edge` and landed with `2.15.0` release.

**Pull request merged**